discovery_engine_core/core/src/stack/filters/article.rs

```diff
@@ -39,52 +36,64 @@ impl ArticleFilter for DuplicateFilter {
     fn apply(
         history: &[HistoricDocument],
         stack: &[Document],
-        articles: Vec<Article>,
+        mut articles: Vec<Article>,
     ) -> Result<Vec<Article>, GenericError> {
-        let urls = history
-            .iter()
-            .map(|doc| doc.url.as_str())
-            .chain(stack.iter().map(|doc| doc.resource.url.as_str()))
-            .collect::<HashSet<_>>();
-
-        let titles = history
-            .iter()
-            .map(|doc| &doc.title)
-            .chain(stack.iter().map(|doc| &doc.resource.title))
-            .collect::<HashSet<_>>();
-
-        let mut articles: BTreeSet<_> = articles.into_iter().map(UniqueArticle).collect();
-
-        articles.retain(|article| {
-            !(urls.contains(&article.0.link.as_str()) || titles.contains(&&article.0.title))
+        // discard dups in the title keeping only the best ranked
+        articles.sort_unstable_by(|art1, art2| match art1.title.cmp(&art2.title) {
+            Ordering::Equal => art1.rank.cmp(&art2.rank),
+            ord => ord,
         });
+        articles.dedup_by(|art1, art2| art1.title == art2.title);
 
-        return Ok(articles.into_iter().map(|u| u.0).collect());
-
-        struct UniqueArticle(Article);
+        // discard dups in the url keeping only the best ranked
+        articles.sort_unstable_by(|art1, art2| match art1.link.cmp(&art2.link) {
+            Ordering::Equal => art1.rank.cmp(&art2.rank),
+            ord => ord,
+        });
+        articles.dedup_by(|art1, art2| art1.link == art2.link);
 
-        impl Eq for UniqueArticle {}
-        impl PartialEq for UniqueArticle {
-            fn eq(&self, other: &Self) -> bool {
-                self.0.link == other.0.link || self.0.title == other.0.title
+        let urls = stack
+            .iter()
+            .map(|doc| (doc.resource.url.as_str(), doc.resource.rank))
+            .chain(history.iter().map(|doc| (doc.url.as_str(), 0)))
```

Comment:
Can't we move the code that remove articles that are already in the history as first pass? and only then dedup and keep the best rank? This should remove the need for the special `0` case.

Reply:
@x3ro yes. you could also maybe think of `0` as `NIL_RANK` here, since historical documents don't have a rank. but yes, there was a particular reason `0` was chosen and not something else.
@acrrd that works too, and perhaps that will be clearer overall to just deal with it separately in a pass.

Reply:
Yeah, I think `NIL_RANK` would still not convey the semantics of the constant, i.e. what using the 0 means there. But I'd be happy with @acrrd's solution also.

Reply:
I think it will be clearer so we don't have to hack in rank values for history values.

So at the light of the latest comments we want:
remove an article if:
 * title or url are the same of one in the history.
 * url is the same of one in the stack documents or batch of articles

Then we want to keep the the article with higher rank if another article or document have same title.

To compare the title we should trim it and put it lowercase.